service/gcs/bridge/bridge.go

```diff
@@ -14,385 +16,563 @@ import (
+			responseBytes, err := json.Marshal(resp.response)
 			if err != nil {
-				logrus.Error(err)
+				responseErrChan <- errors.Wrapf(err, "bridge: failed to marshal JSON for response \"%v\"", resp.response)
+				continue
 			}
-		case prot.ComputeSystemModifySettingsV1:
-			logrus.Info("received from HCS: ComputeSystemModifySettingsV1")
-			response, err = b.modifySettings(message)
-			if err != nil {
-				logrus.Error(err)
+			resp.header.Size = uint32(len(responseBytes) + prot.MessageHeaderSize)
+			if err := binary.Write(b.commandConn, binary.LittleEndian, resp.header); err != nil {
+				responseErrChan <- errors.Wrap(err, "bridge: failed writing message header")
+				continue
 			}
-		default:
-			// TODO: Should a response be returned in this case?
-			logrus.Error(errors.Errorf("received invalid header type code from HCS: 0x%x", header.Type))
-		}
 
-		// Set the error fields on the response if an error was encountered.
-		if err != nil {
-			switch response := response.(type) {
-			case *prot.MessageResponseBase:
-				b.setErrorForResponseBase(response, err)
-			case *prot.ContainerCreateResponse:
-				b.setErrorForResponseBase(response.MessageResponseBase, err)
-			case *prot.ContainerExecuteProcessResponse:
-				b.setErrorForResponseBase(response.MessageResponseBase, err)
-			case *prot.ContainerWaitForProcessResponse:
-				b.setErrorForResponseBase(response.MessageResponseBase, err)
-			case *prot.ContainerGetPropertiesResponse:
-				b.setErrorForResponseBase(response.MessageResponseBase, err)
-			default:
-				// TODO: Should this error be handled better?
-				return errors.Errorf("invalid response type: %T", response)
+			if _, err := b.commandConn.Write(responseBytes); err != nil {
+				responseErrChan <- errors.Wrap(err, "bridge: failed writing message payload")
+				continue
 			}
+			logrus.Infof("bridge: response sent: '%s' to HCS\n", responseBytes)
 		}
+	}()
+	// If we get any errors. We return from Listen and shutdown the bridge connection.
+	// TODO: We should likely just log a generic response and keep listening?
+	select {
+	case conerr = <-requestErrChan:
+		break
+	case conerr = <-responseErrChan:
+		break
+	case <-b.quitChan:
+		break
+	}
+	return conerr
+}
 
-		// Send a response to the HCS, but only if a response was specified.
-		if response != nil {
-			if err := b.sendResponse(response, header); err != nil {
-				return errors.Wrap(err, "failed to send response to HCS")
-			}
-		}
+// PublishNotification writes a specific notification to the bridge.
+func (b *Bridge) PublishNotification(n *prot.ContainerNotification) {
+	if n == nil {
+		panic("bridge: cannot publish nil notification")
+	}
+
+	resp := bridgeResponse{
+		header: &prot.MessageHeader{
+			Type: prot.ComputeSystemNotificationV1,
+			ID:   0,
+		},
+		response: n,
 	}
+	b.responseChan <- resp
 }
 
-func (b *bridge) createContainer(message []byte) (*prot.ContainerCreateResponse, error) {
+func (b *Bridge) createContainer(w ResponseWriter, r *Request) {
```

Comment:
Hmm, OK. I guess I'm less interesting in the bridge/core distinction and more in separating out JSON unmarshaling from the core logic which takes structs without having to worry about how they were produced. So I'm cool with unmarshaling not being in the "bridge", as long as the GCS core methods don't have to perform the unmarshaling themselves (the unmarshaling is done either by the handler or helper functions or whatever). It seems like we're on the same page I think, I was just considering the handlers to be part of the "bridge". :)